command/jsonconfig/config.go

```diff
@@ -358,6 +349,62 @@ func marshalResources(resources map[string]*configs.Resource, schemas *terraform
 	return rs, nil
 }
 
+func marshalDataResources(c *configs.Config, schemas *terraform.Schemas, moduleAddr string) ([]resource, error) {
+	var rs []resource
+	for _, v := range c.Module.DataResources {
```

Comment:
I see that you had to split this into two functions because now one must look at `Module.ManagedResources` and the other at `Module.DataResources`.

However, both of those fields are of the same type and I think ought to have identical processing aside from which map we're looking in, so it seems a shame to have to write essentially the same logic twice and risk them drifting apart later.

Could we potentially add a new function here with the following signature that would contain the main functionality common to both resource modes?

```go
func marshalResource(rc *configs.Resource, moduleAddr addrs.Module, provider addrs.Provider, schemas *terraform.Schemas) (resource, error) {
    // ...
}
```

My hope then would be that this pair of functions could be smaller, both calling into the shared function, e.g.:

```go
    for _, v := range c.Module.DataResources {
        fqn := c.ProviderForLocalConfigAddr(v.ProviderConfigAddr())
        r, err := marshalResource(v, moduleAddr, fqn, schemas)
        if err != nil {
            return nil, err
        }
        rs = append(rs, r)
    }
```

With that said, I'm not sure if there's some other significant difference between managed and data resources I'm not seeing here. If so, never mind!

Reply:
nope, there's no reason not to do a bigger refactor like that, I just opted not to do it right now and instead flagged it as a `todo` because this PR felt pretty unwieldy as is :) I can clean that up, but I'll wait to see what comes of the rest of the PR review (in case I end up reverting a bunch of commits)

Reply:
Ahh, sorry... I guess I missed the todo for all the trees. That's fine... was just trying to make sure I was understanding what was going on here.